### ISSUE TITLE
Improve support for L4 Load Balancers

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc#7957]
TCP-based load balancers are common, in part because they are technology-transparent, but also because HTTP/2 is so new. Users of Kubernetes, nginx, and similar proxies are using L4 today, and experiencing pain. We still suggest using a L7 HTTP/2 proxy like nghttp2, but we may need to consider better support for L4 LBs.

The main things that may be necessary from my perspective are:
1. Cycle through connections more frequently. Basically, after a connection has been used for `X` time, start a new connection to allow the LB to choose a new backend. This is probably best served as server-side configuration.
2. Use multiple connections to the LB on client-side. This is effectively round robin, but with the same host. We normally would only consider this if TLS/HTTP/2 is saturating a core, but this might be another use-case. It's unclear if this is necessary though (based on how imbalanced of load users are experiencing; environment-specific)

**Comment by ejona86** [comment]
CC @louiscryan

**Comment by ejona86** [comment]
I'd love to hear from others if my suggestion would address their L4 woes. It is useful to know whether (1) and/or (2) is what you want.

**Comment by lukaszx0** [comment]
@ejona86 in general, we do both (1) and (2) in our proprietary RPC system and we'd like to have it the same way in our gRPC integration

**Comment by blakeday** [comment]
Agreed with lukaszx0. We would need both (1) and (2) to alleviate the issue entirely.

**Comment by barroca** [comment]
it looks like a similar issue that i'd like to share, maybe we can help each other figuring out what is the issue. I have a docker swarm cluster with 4 servers (2 of each service), one is a http<->grpc proxy in node.js and the other, in python, receives the grpc request.  I facing a similar issue that seems to be something like this, when the servers is idle for a certain time the grcp crashes with 14 on some requests.

**Comment by legendpeng** [comment]
hi, 
  We had this issue and using a managed pool of channels at client side then iterator on each request. While reading latest release notes and according to https://docs.google.com/document/d/1exE0_QNwbHJesD-Opl2D6qX1wujADkUm5OpFykks3C8/edit#

specially this:
L4 load-balancing: now LoadBalancer is free to create multiple Subchannels for the same EAG.

I think #2 has been fixed/improved? however i still don't know what exactly need to be done. any example code?

Thanks!

**Comment by lukaszx0** [comment]
Yeah so with new LBv2 API merged in and promoted as main one you're now able to basically achieve both (1) and (2). In our gRPC in-house integration we have custom `LoadBalancer` implementation which does (2) and could as well do (1) - we didn't implemented (1) for unrelated reasons but it'd be technically possible to do that. So it is possible now to do pretty much everything but it requires a custom `LoadBalancer` implementation - the `RoundRobin` that we ship in core won't be able to do it.

I think we could solve (2) with adding ability to specify some params in `RoundRobin` like `numConnections` which users could use when specifying a `loadBalancerFactory` (currently `RoundRobin` isn't the default one, it's `PickFirst`). This would be relatively easy to implement. It's just a question if we want to grow `RoundRobin` implementation, and since it looks like this will be very common need then maybe it's OK.

For (1) we'd need a new API on the `ManagedChannel`. The idea I have in my head is to have something like `loadBalancerLifetime` or maybe extend, or provide an overload for `loadBalancerFactory` which would let specify the `LoadBalancer.Factory` lifetime. When the lifetime would be reached, we'd reboot `LoadBalancer` by shutting it down and recreating it using factory. This change is a bit more complicated and would require more consideration.

@ejona86 and @zhangkun83 could you chime in? what do you think about my proposed solutions?

**Comment by zhangkun83** [comment]
(1) can be achieved without changing `ManagedChannel` interface. The `LoadBalancer` could periodically close and re-create `Subchannel`s.
For (2), rather than extending the existing `RoundRobin`,  it might be better to create a L4-specific `LoadBalancer` and does both (1) and (2), and put the common logic between `L4` and `RoundRobin` in a base class.

**Comment by ejona86** [comment]
It is possible to implement both (1) and (2) in Java, but that would still be short-term. We're working on features that you just need to configure to be on.

(1) is being solved with a "max age" on the server-side. Work is already underway for this in Java. The other languages should be pretty close as well (I don't know if they are further along or not).

We're still figuring out the design for (2). With the new Java LBv2 API we can do this at several layers: 1) at the Channel layer by wrapping the Channel, 2) in the LB, 3) in the name resolver, and 4) in the InternalSubchannel (this is a Java concept; there is something similar in other languages). We're figuring out where. Once we have a design, I imagine the implementation will be fairly easy.

If you need a fix for (2) _today_ in Java, I'd suggest the Channel layer, with something like:
```java
public final class MultiChannel extends Channel {
  private final ImmutableList<Channel> channels;
  private final AtomicInteger pos = new AtomicInteger();

  public MultiChannel(List<Channel> channels) {
    this.channels = ImmutableList.copyOf(channels);
  }

  public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(
      MethodDescriptor<ReqT, RespT> m, CallOptions o) {
    int idx = UnsignedInts.remainder(pos.getAndIncrement(), channels.size());
    return channels.get(idx).newCall(m, o);
  }

  public String authority() {
    return channels.get(0).authority();
  }
}
```

Note that you could pass in a ManagedChannelBuilder and have the code create channels as necessary. (1) could be solved in a similar way. Only people needing this in many different applications should touch the LB solution, since it is more complex to implement but it probably easier to enable.

**Comment by ejona86** [comment]
I have opened [gRFC A9](https://github.com/grpc/proposal/pull/23) for "max age" which addresses (1).

**Comment by markdroth** [comment]
Note that item (2) is being tracked in #11704.

**Comment by markdroth** [comment]
#11704 got closed, but the same issue is being tracked in #21386.

**Comment by markdroth** [comment]
I think both of these cases are addressed now.
